### PR TITLE
Refactor `Function`

### PR DESCRIPTION
**Pull request by HalidOdat**
This PR is a preparation for the function design discussed in #445 

It changes the following:
 - Moved `Function::call()` => `GcObject::call()` this is done because we need the object that is being called to be garbage collected, because we put it in the environment. This was awkward before since we pass a clone of object and the access the `Function` and call it and pass the function object. 
 - Moved `Function::construct()` => `GcObject::construct()` same as `Function::call()`
 - Made `Function` be an enum since many fields that are for ordinary functions where there for builtins.
 - Removed `InternalState` since we will replace it with a `NativeObject`.
 - Put `RegExp` in box because it was really big (136 bytes)
 - Added `Interpreter::create_builtin_function(name, length, function pointer) -> GcObject`
 - Added `Interpreter::register_global_function(name, length, function pointer)`
 - Shrink `Function` from `56` bytes to `48` bytes
 - Shrink `ObjectData` from `144` bytes to `80` bytes
 - Shrink `Object` from `272` bytes to `192` bytes

### CODE REVIEW

**Comment by Razican** [comment]
If I'm not mistaken, this would de facto close #387 and give a nice solution to https://github.com/boa-dev/boa/issues/363#issuecomment-626170196, right?

**Comment by HalidOdat** [comment]
> If I'm not mistaken, this would de facto close #387 and give a nice solution to [#363 (comment)](https://github.com/boa-dev/boa/issues/363#issuecomment-626170196), right?

I guess so, forgot there was an PR to fix it. This PR only removes it but there will also be PR to add `NativeObject` after some discussion. I have a proof of concept locally. I'll create a PR to see and discuss on it.

**Comment by jasonwilliams** [comment]
Looks good so far from what i can see

**Comment by HalidOdat** [comment]
> Im guessing this and the NativeObject will need to go in together?

No. Not necessarily they accomplish different goals. This PR does some prepwork for the new function design, and #627 (`NativeObject`) introduces the native objects

File: boa/src/builtins/function/mod.rs [hunk 1]
```diff
@@ -18,84 +18,43 @@ use crate::{
...
     syntax::ast::node::{FormalParameter, StatementList},
     BoaProfiler, Result,
 };
 use bitflags::bitflags;
 use gc::{unsafe_empty_trace, Finalize, Trace};
 use std::fmt::{self, Debug};
 
-/// _fn(this, arguments, ctx) -> Result<Value>_ - The signature of a built-in function
-pub type NativeFunctionData = fn(&Value, &[Value], &mut Interpreter) -> Result<Value>;
+/// _fn(this, arguments, ctx) -> ResultValue_ - The signature of a built-in function
```
Comment (Razican) [inline boa/src/builtins/function/mod.rs:30]: I think this comment needs to be updated back again.
```suggestion
/// _fn(this, arguments, ctx) -> Result<Value>_ - The signature of a built-in function
```

**Pull request merged**